pdns/pdns_recursor.cc

```diff
@@ -797,38 +798,34 @@ static void protobufLogQuery(uint8_t maskV4, uint8_t maskV6, const boost::uuids:
   Netmask requestorNM(remote, remote.sin4.sin_family == AF_INET ? maskV4 : maskV6);
   ComboAddress requestor = requestorNM.getMaskedNetwork();
   requestor.setPort(remote.getPort());
-  RecProtoBufMessage message(DNSProtoBufMessage::Query, uniqueId, &requestor, &local, qname, qtype, qclass, id, tcp, len);
-  message.setServerIdentity(SyncRes::s_serverID);
-  message.setEDNSSubnet(ednssubnet, ednssubnet.isIPv4() ? maskV4 : maskV6);
-  message.setRequestorId(requestorId);
-  message.setDeviceId(deviceId);
-  message.setDeviceName(deviceName);
+
+  pdns::ProtoZero::Message m{128, std::string::size_type(policyTags.empty() ? 0 : 64)}; // XXX guesses at sizes
```

Comment:
Did you consider a `pdns::ProtoZero::Message` constructor taking a reference to existing string objects instead? That way I think we could pass already allocated strings, like `thread_local` ones, which would allow us to reuse the same allocations over and over.
That won't work in `startDoResolve()`, of course, because we can switch to a different `mthread` from that function.

Reply:
I considered it, but wanted to avoid the extra complexity at the moment. We can always optimize this later.